### PR TITLE
Add copy code to clipboard button for `rustdoc`

### PR DESCRIPTION
**Pull request by grey4owl**
A very simple button that was missing in the upper right corner to copy the code blocks to the clipboard. It is hidden by default, but when you hover over the code block (`<pre>`) element then it appears.. It also works on mobile devices and it will appears after you tap on the code block.

It looks like this:
Screenshot from 2024-01-13 17-11-48
Screenshot from 2024-01-13 17-12-02

### CODE REVIEW

File: src/librustdoc/html/static/js/main.js [hunk 1]
```diff
@@ -1734,6 +1734,37 @@ href="https://doc.rust-lang.org/${channel}/rustdoc/read-documentation/search.htm
     resizer.addEventListener("pointerdown", initResize, false);
 }());
 
+// This section handles the copy buttons that appears in top right corner of the code blocks
+(function() {
+    let preElements = document.querySelectorAll("pre:has(code)");
```
Comment (fmease) [inline src/librustdoc/html/static/js/main.js:1739]: `:has` is not supported on Firefox Mobile: https://caniuse.com/css-has

File: src/librustdoc/html/static/js/main.js [hunk 2]
```diff
@@ -1734,6 +1734,37 @@ href="https://doc.rust-lang.org/${channel}/rustdoc/read-documentation/search.htm
...
+    preElements.forEach(function(pre, index) {
+        let resetTimeout = null;
+        let copyBtn = pre.querySelector("button[class='copy-code']");
```
Comment (fmease) [inline src/librustdoc/html/static/js/main.js:1742]:
```suggestion
        let copyBtn = pre.querySelector("button.copy-code");
```

File: src/librustdoc/html/static/js/main.js [hunk 3]
```diff
@@ -1734,6 +1734,37 @@ href="https://doc.rust-lang.org/${channel}/rustdoc/read-documentation/search.htm
...
+        let icon = copyBtn.innerHTML;
+        pre.addEventListener("mouseenter", function() {
+            copyBtn.style.opacity = "1";
```
Comment (fmease) [inline src/librustdoc/html/static/js/main.js:1745]: You should be able to express this in pure CSS with:

```
pre:hover > button.copy-code { opacity: 1; }
```

**Review by fmease — COMMENTED** [review]
Thanks for working on this, I've only skimmed your PR so far and have a couple of comments. Could you please squash the 8 commits into one (or at least fewer) and `./x fmt` your changes (that's why CI has failed).

**Comment by grey4owl** [comment]
@fmease I messed up some things, Im  closing this PR and I'll make a new one. Sorry .. :smiling_face_with_tear:

**Comment by fmease** [comment]
All good, now worries! You can write `r? @fmease` in the description of your new PR and it will automatically assign me.

**Comment by egtwp** [comment]
@grey4owl had you opened a new MR about this? Thank u!

**Comment by fmease** [comment]
They did, yes. Namely #119979 which was closed, too, due inactivity/time constraints of the author. Feel free to take over the work of #119979!

**Comment by GuillaumeGomez** [comment]
A few things to take into account for this feature:

 * Considering it's only working when JS is enabled, it should not be rendered into HTML files but instead generated with JS.
 * To prevent this generation to impact load time too much, it should be done in two steps:
   1. Add a `onhover` event on all rust codeblocks.
   2. When this event is triggered, generate the button and remove the event handler on the item.

**Comment by GuillaumeGomez** [comment]
I implemented the feature. Just needs some UI tests and then I'll send a PR.

**Comment by GuillaumeGomez** [comment]
Forgot to mention but PR is open: https://github.com/rust-lang/rust/pull/125779

**Pull request closed without merging**